File: src/main/java/me/syari/ss/discord/internal/entities/TextChannel.java
package me.syari.ss.discord.internal.entities;

import me.syari.ss.discord.api.JDA;
import me.syari.ss.discord.api.entities.ISnowflake;
import me.syari.ss.discord.internal.JDAImpl;
import me.syari.ss.discord.internal.requests.Route;
import me.syari.ss.discord.internal.requests.restaction.MessageAction;
import me.syari.ss.discord.internal.utils.Checks;
import me.syari.ss.discord.internal.utils.cache.SnowflakeReference;
import org.jetbrains.annotations.NotNull;

public class TextChannel implements ISnowflake, Comparable<TextChannel> {
    protected final long id;
    protected final SnowflakeReference<Guild> guild;
    protected final JDAImpl api;
    protected String name;

    public TextChannel(long id, @NotNull Guild guild) {
        this.id = id;
        this.api = guild.getJDA();
        this.guild = new SnowflakeReference<>(guild, api::getGuildById);
    }

    public void setName(String name) {
        this.name = name;
    }

    @NotNull
    public String getName() {
        return name;
    }

    @NotNull
    public Guild getGuild() {
        return guild.resolve();
    }

    @Override
    public long getIdLong() {
        return id;
    }

    @NotNull
    public String getAsMention() {
        return "<#" + id + '>';
    }

    @NotNull
    public JDA getJDA() {
        return api;
    }

    @NotNull
    public MessageAction sendMessage(@NotNull CharSequence text) {
        Checks.notEmpty(text, "Provided text for message");
        Checks.check(text.length() <= 2000, "Provided text for message must be less than 2000 characters in length");

        Route.CompiledRoute route = Route.Messages.SEND_MESSAGE.compile(getId());
        if (text instanceof StringBuilder) {
            return new MessageAction(getJDA(), route, this, (StringBuilder) text);
        } else {
            return new MessageAction(getJDA(), route, this).append(text);
        }
    }

    @Override
    public String toString() {
        return "TextChannel:" + getName() + '(' + id + ')';
    }

    @Override
    public int compareTo(@NotNull TextChannel channel) {
        return Long.compareUnsigned(id, channel.getIdLong());
    }

    @Override
    public int hashCode() {
        return Long.hashCode(id);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this)
            return true;
        if (!(obj instanceof TextChannel))
            return false;
        TextChannel channel = (TextChannel) obj;
        return channel.getIdLong() == getIdLong();
    }
}

File: src/main/kotlin/me/syari/ss/discord/internal/entities/TextChannel.kt
package me.syari.ss.discord.internal.entities

import me.syari.ss.discord.api.data.DataContainer
import me.syari.ss.discord.api.requests.Request
import me.syari.ss.discord.api.requests.Response
import me.syari.ss.discord.internal.requests.Requester
import me.syari.ss.discord.internal.requests.RestAction
import me.syari.ss.discord.internal.requests.Route
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.lang.Long.compareUnsigned
import java.lang.Long.hashCode

class TextChannel(
    override val idLong: Long, val guild: Guild, val name: String
): WithId, Comparable<TextChannel> {
    companion object {
        private const val MAX_CONTENT_LENGTH = 2000

        private val textChannelCache = mutableMapOf<Long, TextChannel?>()

        fun get(id: Long): TextChannel? {
            return textChannelCache.getOrPut(id) { getFromGuild(id) }
        }

        private fun getFromGuild(id: Long): TextChannel? {
            for (guild in Guild.allGuild) {
                return guild.getTextChannel(id) ?: continue
            }
            return null
        }
    }

    val asMention: String
        get() = "<#$idLong>"

    fun sendMessage(text: String) {
        val length = text.length
        if (length == 0) return
        if (MAX_CONTENT_LENGTH < length) {
            sendMessage(text.substring(0, 2000))
            sendMessage(text.substring(2000))
            return
        }
        val route = Route.sendMessageRoute(id)
        val messageAction = MessageAction(route, this, text)
        messageAction.queue()
    }

    override fun toString(): String {
        return "TextChannel:$name($idLong)"
    }

    override fun compareTo(other: TextChannel): Int {
        return compareUnsigned(idLong, other.idLong)
    }

    override fun hashCode(): Int {
        return hashCode(idLong)
    }

    override fun equals(other: Any?): Boolean {
        if (other === this) return true
        if (other !is TextChannel) return false
        return other.idLong == idLong
    }

    private class MessageAction(
        route: Route, private val channel: TextChannel, private val content: String
    ): RestAction<Message>(route) {
        override fun finalizeData(): RequestBody {
            return DataContainer().apply {
                put("content", content)
            }.toString().toRequestBody(Requester.MEDIA_TYPE_JSON)
        }

        override fun handleSuccess(response: Response, request: Request<Message>) {
            val message = EntityBuilder.createMessage(response.dataObject, channel)
            request.onSuccess(message)
        }
    }
}